### PR TITLE
layer auto fusion for dnn graph

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

add auto fusion for dnn graph

## How was this patch tested?

unit test and example test by hand

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ */
+
+package com.intel.analytics.bigdl.nn.mkldnn
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.Node
+
+private[mkldnn] object Fusion {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: add class comments?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+  private val fuse = System.getProperty("bigdl.mkldnn.fusion", "false").toBoolean
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: I assume fusion normally always benefit inference,  in what case we may not want to do fusion?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      case _ =>
+    }
+  }
+
+  /**
+   * fuse conv(without relu or bn fusion) with bn
+   * if bn has fused with relu, then fuse relu and bn with conv
+   * @param node
+   */
+  private def fusionBn(node: Node[AbstractModule[Activity, Activity, Float]]): Unit = {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: Bn -> BN ?  for convention

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/Fusion.scala]: Maybe we can set it default to true after intergating with int8

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/377/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 4]
```diff
@@ -390,10 +390,25 @@ class DnnGraph(
     }
   }
 
+  // fuse some layers when doing inference
+  private def fusion(): Unit = {
+    if (!this.train) {
+      for (j <- 0 to 1) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: add comment why we do 2 times iteration

**Comment by wzhongyuan** [comment]
please run a Jenkins job and merge it after it passes the test

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/379

**Pull request merged**